From b0e0a9747cb2df84744539f5940d7eda06f2601c Mon Sep 17 00:00:00 2001
From: Lu Jialin <lujialin4@huawei.com>
Date: Mon, 14 Nov 2022 11:14:02 +0800
Subject: [PATCH 0072/1398] [Huawei] memcg: Introduce
 CONFIG_MEMCG_V1_THRESHOLD_QOS

Offering: HULK
hulk inclusion
category: feature
bugzilla: 187972

-------------------------------

Introduce CONFIG_MEMCG_V1_THRESHOLD_QOS to isolate memcg qos management
feature from baseline.

Signed-off-by: Lu Jialin <lujialin4@huawei.com>
---
 arch/arm64/configs/openeuler_defconfig | 1 +
 arch/x86/configs/openeuler_defconfig   | 1 +
 include/linux/memcontrol.h             | 7 ++++++-
 init/Kconfig                           | 5 +++++
 mm/memcontrol.c                        | 6 ++++++
 5 files changed, 19 insertions(+), 1 deletion(-)

diff --git a/arch/arm64/configs/openeuler_defconfig b/arch/arm64/configs/openeuler_defconfig
index ad37376e4b05..e5558806da32 100644
--- a/arch/arm64/configs/openeuler_defconfig
+++ b/arch/arm64/configs/openeuler_defconfig
@@ -132,6 +132,7 @@ CONFIG_CGROUPS=y
 CONFIG_CGROUP_V1_KILL=y
 CONFIG_PAGE_COUNTER=y
 CONFIG_MEMCG=y
+CONFIG_MEMCG_V1_THRESHOLD_QOS=y
 CONFIG_MEMCG_SWAP=y
 CONFIG_MEMCG_KMEM=y
 CONFIG_MEMCG_MEMFS_INFO=y
diff --git a/arch/x86/configs/openeuler_defconfig b/arch/x86/configs/openeuler_defconfig
index e2fc38353a49..da7549b05181 100644
--- a/arch/x86/configs/openeuler_defconfig
+++ b/arch/x86/configs/openeuler_defconfig
@@ -151,6 +151,7 @@ CONFIG_CGROUPS=y
 CONFIG_CGROUP_V1_KILL=y
 CONFIG_PAGE_COUNTER=y
 CONFIG_MEMCG=y
+CONFIG_MEMCG_V1_THRESHOLD_QOS=y
 CONFIG_MEMCG_SWAP=y
 CONFIG_MEMCG_KMEM=y
 CONFIG_MEMCG_MEMFS_INFO=y
diff --git a/include/linux/memcontrol.h b/include/linux/memcontrol.h
index bf29bd0ffda3..e73ae787e4e2 100644
--- a/include/linux/memcontrol.h
+++ b/include/linux/memcontrol.h
@@ -1212,8 +1212,10 @@ static bool memcg_event_add(struct mem_cgroup *memcg,
 	if (!mem_cgroup_is_root(memcg))
 		return true;
 
+#ifdef CONFIG_MEMCG_V1_THRESHOLD_QOS
 	if (event == MEMCG_OOM_KILL && !cgroup_subsys_on_dfl(memory_cgrp_subsys))
 		return true;
+#endif
 
 	return false;
 }
@@ -1234,7 +1236,10 @@ static inline void memcg_memory_event(struct mem_cgroup *memcg,
 			cgroup_file_notify(&memcg->swap_events_file);
 		else
 			cgroup_file_notify(&memcg->events_file);
-
+#ifndef CONFIG_MEMCG_V1_THRESHOLD_QOS
+		if (!cgroup_subsys_on_dfl(memory_cgrp_subsys))
+			break;
+#endif
 		if (cgrp_dfl_root.flags & CGRP_ROOT_MEMORY_LOCAL_EVENTS)
 			break;
 	} while ((memcg = parent_mem_cgroup(memcg)) &&
diff --git a/init/Kconfig b/init/Kconfig
index 73334071fa32..c4767a427cb5 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -945,6 +945,11 @@ config MEMCG
 	help
 	  Provides control over the memory footprint of tasks in a cgroup.
 
+config MEMCG_V1_THRESHOLD_QOS
+	bool "Qos memcg threshold in v1"
+	depends on MEMCG
+	default n
+
 config MEMCG_SWAP
 	bool
 	depends on MEMCG && SWAP
diff --git a/mm/memcontrol.c b/mm/memcontrol.c
index cc7368380f69..dc88b728288a 100644
--- a/mm/memcontrol.c
+++ b/mm/memcontrol.c
@@ -4546,8 +4546,10 @@ static int mem_cgroup_oom_control_read(struct seq_file *sf, void *v)
 	seq_printf(sf, "under_oom %d\n", (bool)memcg->under_oom);
 	seq_printf(sf, "oom_kill %lu\n",
 		   atomic_long_read(&memcg->memory_events[MEMCG_OOM_KILL]));
+#ifdef CONFIG_MEMCG_V1_THRESHOLD_QOS
 	seq_printf(sf, "oom_kill_local %lu\n",
 		   atomic_long_read(&memcg->memory_events_local[MEMCG_OOM_KILL]));
+#endif
 
 	return 0;
 }
@@ -5109,6 +5111,7 @@ static ssize_t memory_high_write(struct kernfs_open_file *of,
 	return nbytes;
 }
 
+#ifdef CONFIG_MEMCG_V1_THRESHOLD_QOS
 static void __memcg_events_show(struct seq_file *m, atomic_long_t *events)
 {
 	seq_printf(m, "low %lu\n", atomic_long_read(&events[MEMCG_LOW]));
@@ -5133,6 +5136,7 @@ static int memcg_events_local_show(struct seq_file *m, void *v)
 	__memcg_events_show(m, memcg->memory_events_local);
 	return 0;
 }
+#endif
 
 static ssize_t memory_reclaim(struct kernfs_open_file *of, char *buf,
 			      size_t nbytes, loff_t off)
@@ -5410,6 +5414,7 @@ static struct cftype mem_cgroup_legacy_files[] = {
 		.write = mem_cgroup_reset,
 		.read_u64 = mem_cgroup_read_u64,
 	},
+#ifdef CONFIG_MEMCG_V1_THRESHOLD_QOS
 	{
 		.name = "min",
 		.flags = CFTYPE_NOT_ON_ROOT,
@@ -5440,6 +5445,7 @@ static struct cftype mem_cgroup_legacy_files[] = {
 		.file_offset = offsetof(struct mem_cgroup, events_local_file),
 		.seq_show = memcg_events_local_show,
 	},
+#endif
 
 #ifdef CONFIG_MEMCG_RECLAIM
 	{
-- 
2.17.1

